pkg/registrymirror/registrymirrorsstore.go

```diff
@@ -0,0 +1,325 @@
+	// The returned slice may (or may not) include the source image based on the registries config.
+	// Source image must be fully qualified (include the registry hostname).
+	PullSources(srcImage string) ([]string, error)
+}
+
+// FileStore stores/reads the consolidated registries config from a filesystem.
+type FileStore struct {
+	configPath string
+
+	// holds the mirror sets used for managing registry mirrors.
+	icspRules   map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy
+	idmsRules   map[types.UID]*configV1.ImageDigestMirrorSet
+	itmsRules   map[types.UID]*configV1.ImageTagMirrorSet
+	ruleRWMutex sync.RWMutex
+
+	// timer is used when performing delayed writes to filesystem.
+	timer       *time.Timer
+	timerMutex  sync.Mutex
+	updateDelay time.Duration
+
+	systemContext *ciTypes.SystemContext
+}
+
+var _ Store = (*FileStore)(nil)
+
+// fileStoreOption is used to provide functional options to the fileStore.
+type fileStoreOption func(*FileStore)
+
+// WithConfigPath sets the path to read/write the consolidated registries config.
+func WithConfigPath(path string) fileStoreOption {
+	return func(s *FileStore) { s.configPath = path }
+}
+
+// WithDelay sets a delay before the updated config is written to disk.
+func WithDelay(delay time.Duration) fileStoreOption {
+	return func(s *FileStore) { s.updateDelay = delay }
+}
+
+// NewFileStore creates a new FileStore.
+func NewFileStore(opts ...fileStoreOption) *FileStore {
+	s := &FileStore{
+		configPath:  defaultRegistriesPath,
+		updateDelay: defaultDelay,
+		icspRules:   make(map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy),
+		idmsRules:   make(map[types.UID]*configV1.ImageDigestMirrorSet),
+		itmsRules:   make(map[types.UID]*configV1.ImageTagMirrorSet),
+	}
+
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	s.systemContext = &ciTypes.SystemContext{SystemRegistriesConfPath: s.configPath}
+	return s
+}
+
+// Cleanup resets the store which includes in-memory and disk resources.
+func (s *FileStore) Cleanup() {
+	s.ruleRWMutex.Lock()
+	defer s.ruleRWMutex.Unlock()
+
+	s.icspRules = make(map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy)
+	s.idmsRules = make(map[types.UID]*configV1.ImageDigestMirrorSet)
+	s.itmsRules = make(map[types.UID]*configV1.ImageTagMirrorSet)
+
+	if err := os.Remove(s.configPath); err != nil && !os.IsNotExist(err) {
+		log.Warnf("Failed to cleanup registries config at %q: %v", s.configPath, err)
+	}
+	sysregistriesv2.InvalidateCache()
+}
+
+func (s *FileStore) updateConfig() error {
+	if s.updateDelay == 0 {
+		return s.updateConfigNow()
+	}
+
+	return s.updateConfigDelayed()
+}
+
+func (s *FileStore) updateConfigDelayed() error {
+	// Using a different mutex to protect timer vs. rules to avoid deadlock.
+	s.timerMutex.Lock()
+	defer s.timerMutex.Unlock()
+
+	if s.timer != nil {
+		s.timer.Stop()
+	}
+
+	s.timer = time.AfterFunc(s.updateDelay, func() {
```

Comment:
this is optional: as an alternative to use the `time.AfterFunc` and including the mutex for this and having to store the timer, you could achieve the same by using a combination of `concurrency.Signal, concurrency.AfterFunc`.

Essentially, this would be:
```go

var cancelUpdate concurrency.Signal

func doAsync() {
        // Unconditionally signal and then reset the signal in a non-triggered state.
        cancelUpdate.Signal()
        cancelUpdate.Reset()

        concurrency.AfterFunc(delay, func() {
              // do the async update
        }, cancelUpdate)

        return nil
}

```

Reply:
I'll test this out again, I did initially attempt to implement this with `concurrency.AfterFunc` but was fighting with the API so fell back to the standard lib.